src/suncontrol/i/suncontrol_i.c

```diff
@@ -0,0 +1,192 @@
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <suncontrol/suncontrol_i.h>
+#include <sundials/sundials_math.h>
+
+
+/* ---------------
+ * Macro accessors
+ * --------------- */
+
+#define SC_I_CONTENT(C)     ( (SUNControlContent_I)(C->content) )
+#define SC_I_K1(C)          ( SC_I_CONTENT(C)->k1 )
+#define SC_I_BIAS(C)        ( SC_I_CONTENT(C)->bias )
+#define SC_I_P(C)           ( SC_I_CONTENT(C)->p )
+#define SC_I_PQ(C)          ( SC_I_CONTENT(C)->pq )
+
+/* ------------------
+ * Default parameters
+ * ------------------ */
+
+#define DEFAULT_K1     RCONST(1.0)
+#define DEFAULT_BIAS   RCONST(1.5)
+#define DEFAULT_PQ     SUNFALSE
+#define TINY           RCONST(1.0e-10)
+
+
+/* -----------------------------------------------------------------
+ * exported functions
+ * ----------------------------------------------------------------- */
+
+/* -----------------------------------------------------------------
+ * Function to create a new I controller
+ */
+
+SUNControl SUNControlI(SUNContext sunctx)
+{
+  SUNControl C;
+  SUNControlContent_I content;
+
+  /* Create an empty controller object */
+  C = NULL;
+  C = SUNControl_NewEmpty(sunctx);
+  if (C == NULL) { return (NULL); }
+
+  /* Attach operations */
+  C->ops->gettype           = SUNControlGetType_I;
+  C->ops->estimatestep      = SUNControlEstimateStep_I;
+  C->ops->setdefaults       = SUNControlSetDefaults_I;
+  C->ops->write             = SUNControlWrite_I;
+  C->ops->setmethodorder    = SUNControlSetMethodOrder_I;
+  C->ops->setembeddingorder = SUNControlSetEmbeddingOrder_I;
+  C->ops->seterrorbias      = SUNControlSetErrorBias_I;
+  C->ops->space             = SUNControlSpace_I;
+
+  /* Create content */
+  content = NULL;
+  content = (SUNControlContent_I)malloc(sizeof *content);
+  if (content == NULL)
+  {
+    (void) SUNControl_Destroy(C);
+    return (NULL);
+  }
+
+  /* Attach content */
+  C->content = content;
+
+  /* Initialize method order */
+  content->p = 1;
+
+  /* Fill content with default values */
+  SUNControlSetDefaults_I(C);
+
+  return (C);
+}
+
+/* -----------------------------------------------------------------
+ * Function to set I parameters
+ */
+
+int SUNControlI_SetParams(SUNControl C, sunbooleantype pq,
+                          realtype k1)
+{
+  /* store legal inputs, and return with success */
+  SC_I_PQ(C) = pq;
+  if (k1 >= RCONST(0.0)) { SC_I_K1(C) = k1; }
+  return SUNCONTROL_SUCCESS;
+}
+
+
+
+/* -----------------------------------------------------------------
+ * implementation of controller operations
+ * ----------------------------------------------------------------- */
+
+SUNControl_Type SUNControlGetType_I(SUNControl C) { return SUNDIALS_CONTROL_H; }
+
+int SUNControlEstimateStep_I(SUNControl C, realtype h,
+                               realtype dsm, realtype* hnew)
+{
+  /* set usable time-step adaptivity parameters */
+  const realtype k1 = -SC_I_K1(C) / SC_I_P(C);
```

Comment:
If we edit the controller struct to include p and q, the denominator should be `min(p,q)+1`

Reply:
I'm good with most of these suggestions, but I have one follow-on question.  Suppose a user knows that their problem will exhibit severe order reduction but they want to use a higher-order DIRK method anyways, and in so doing they want to modify the controller to use a smaller exponent to account for the reduced method/embedding order.  

I see no way to do this (or equivalently, to allow current ARKODE users to get the same exponent they have now), based on these suggestions.  Even if a user manually called your proposed `SUNAdaptController_SetOrder` routine to set their value, it would be overridden when `arkStep_Init` calls that same routine.  

What about if the controller had a corresponding "GetOrder" routine, under the restriction that controllers with non-NULL `SetOrder` routine should initialize their order to 0.  Then if a user never called `SetOrder` directly, ARKStep could call `GetOrder` to see if this was nonzero before itself callin `SetOrder`?

Reply:
In that case, the user could create a tableau with `ARKodeButcherTable_Create` and set p and q to values that account for order reduction. Unfortunately, I don't see a straightforward way to load something from our catalog and modify the order since we don't have getters or setters for table properties. This is a much simpler solution but more verbose and manual.

Your solution also seems reasonable, but could have strange side effects if the same controller is used for two integrations with methods of different order.

Reply:
Historically, whenever we have changed any defaults, we always strive for a "simple" way for users to recover what they had before (case in point, consider Carol's current collaborators who get significantly fewer steps with the current off-by-one exponents).  

While I like your idea of manipulating the Butcher table structure, I don't see this as "simple" for users.  I'll give some more thought to this question today (I teach and have office hours all afternoon), and come up with something tonight/tomorrow.

Reply:
I think that the simplest solution is to add an `AdjustOrder` routine to our provided controllers, where the integer argument is then applied to whatever order it is subsequently handed by the integrator.  The integer order adjustment can then always be undone by a second call with a `0` input.

I'll work on implementing Steven's recommendations, plus the `AdjustOrder` routine, tomorrow.

Reply:
The `AdjustOrder` solution sounds good